File: python/agentic_memory/store/file.py
"""File-based storage backend. Persists to JSON on disk."""

from __future__ import annotations

import json
import os
from typing import Optional

from ..types import MemoryEntry, MemoryType, ImportanceLevel, RetrievalQuery


class FileStore:
    """
    File-based storage backend. Persists to a JSON file on disk.
    Good for single-process agents that need persistence across restarts.
    Not suitable for concurrent multi-agent access.
    """

    def __init__(self, file_path: str) -> None:
        self._file_path = file_path
        self._entries: dict[str, MemoryEntry] = {}
        self._dirty = False
        self._load()

    def _load(self) -> None:
        if not os.path.exists(self._file_path):
            return
        try:
            with open(self._file_path, "r", encoding="utf-8") as f:
                data = json.load(f)
            for item in data:
                item["type"] = MemoryType(item["type"])
                item["importance"] = ImportanceLevel(item["importance"])
                entry = MemoryEntry(**item)
                self._entries[entry.id] = entry
        except (json.JSONDecodeError, KeyError, TypeError):
            self._entries.clear()

    def _persist(self) -> None:
        if not self._dirty:
            return
        dir_path = os.path.dirname(self._file_path)
        if dir_path and not os.path.exists(dir_path):
            os.makedirs(dir_path, exist_ok=True)

        data = []
        for entry in self._entries.values():
            d = vars(entry).copy()
            d["type"] = d["type"].value
            d["importance"] = d["importance"].value
            data.append(d)

        with open(self._file_path, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=2)
        self._dirty = False

    async def get(self, id: str) -> Optional[MemoryEntry]:
        entry = self._entries.get(id)
        return entry if entry is None else MemoryEntry(**vars(entry))

    async def get_all(self, scope: Optional[str] = None) -> list[MemoryEntry]:
        entries = list(self._entries.values())
        if scope is not None:
            entries = [e for e in entries if e.scope == scope]
        return entries

    async def set(self, entry: MemoryEntry) -> None:
        self._entries[entry.id] = MemoryEntry(**vars(entry))
        self._dirty = True
        self._persist()

    async def delete(self, id: str) -> bool:
        if id in self._entries:
            del self._entries[id]
            self._dirty = True
            self._persist()
            return True
        return False

    async def clear(self, scope: Optional[str] = None) -> None:
        if scope is None:
            self._entries.clear()
        else:
            to_delete = [k for k, v in self._entries.items() if v.scope == scope]
            for k in to_delete:
                del self._entries[k]
        self._dirty = True
        self._persist()

    async def search(self, query: RetrievalQuery) -> list[MemoryEntry]:
        results = list(self._entries.values())
        if query.scope is not None:
            results = [e for e in results if e.scope == query.scope]
        if query.types:
            results = [e for e in results if e.type in query.types]
        return results

    @property
    def size(self) -> int:
        return len(self._entries)

File: src/store/file.ts
import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'fs';
import { join, dirname } from 'path';
import type { StorageBackend, MemoryEntry, RetrievalQuery } from '../core/types.js';

/**
 * File-based storage backend. Persists to a JSON file on disk.
 * Good for single-process agents that need persistence across restarts.
 * Not suitable for concurrent multi-agent access - use Redis/Postgres for that.
 */
export class FileStore implements StorageBackend {
  private entries: Map<string, MemoryEntry> = new Map();
  private filePath: string;
  private dirty = false;

  constructor(filePath: string) {
    this.filePath = filePath;
    this.load();
  }

  private load(): void {
    if (existsSync(this.filePath)) {
      try {
        const raw = readFileSync(this.filePath, 'utf-8');
        const data: MemoryEntry[] = JSON.parse(raw);
        for (const entry of data) {
          this.entries.set(entry.id, entry);
        }
      } catch {
        // Corrupted file - start fresh
        this.entries.clear();
      }
    }
  }

  private persist(): void {
    if (!this.dirty) return;
    const dir = dirname(this.filePath);
    if (!existsSync(dir)) {
      mkdirSync(dir, { recursive: true });
    }
    const data = Array.from(this.entries.values());
    writeFileSync(this.filePath, JSON.stringify(data, null, 2), 'utf-8');
    this.dirty = false;
  }

  async get(id: string): Promise<MemoryEntry | null> {
    return this.entries.get(id) ?? null;
  }

  async getAll(scope?: string): Promise<MemoryEntry[]> {
    const all = Array.from(this.entries.values());
    if (!scope) return all;
    return all.filter(e => e.scope === scope);
  }

  async set(entry: MemoryEntry): Promise<void> {
    this.entries.set(entry.id, { ...entry });
    this.dirty = true;
    this.persist();
  }

  async delete(id: string): Promise<boolean> {
    const deleted = this.entries.delete(id);
    if (deleted) {
      this.dirty = true;
      this.persist();
    }
    return deleted;
  }

  async clear(scope?: string): Promise<void> {
    if (!scope) {
      this.entries.clear();
    } else {
      for (const [id, entry] of this.entries) {
        if (entry.scope === scope) {
          this.entries.delete(id);
        }
      }
    }
    this.dirty = true;
    this.persist();
  }

  async search(query: RetrievalQuery): Promise<MemoryEntry[]> {
    let results = Array.from(this.entries.values());
    if (query.scope) {
      results = results.filter(e => e.scope === query.scope);
    }
    if (query.types && query.types.length > 0) {
      results = results.filter(e => query.types!.includes(e.type));
    }
    return results;
  }

  get size(): number {
    return this.entries.size;
  }
}
